### ISSUE TITLE
GHA workflow R&D

### ISSUE THREAD

**Issue by nilsnolde** [gis-ops/py-osrm#11]
Having CI workflows running for PR & e.g. distribution for Linux, Mac, Win poses some challenges (esp for the latter two).. To test the code on every push or package it, we'd need to:
- build all OSRM dependencies on Mac & Win (Linux is easier here)
- build all OSRM binaries so we can bind to them

Ideally we only (re-)build dependencies when they actually need to be re-built (i.e. upstream needs new minimum versions). And ideally we only (re-)build OSRM binaries when we actually upgrade the OSRM version.

IMHO we should investigate if we can't make this more optimal and have workflows which build dependencies & OSRM when we **need it** and push them to some remote storage where our binding workflows can pull from. I didn't really look into this too much, but first thoughts:
- Github has artifacts which can be pushed & pulled. Since a few months they're also exposed via Github's HTTP API, so pushing/pulling to/from Github artifacts could be scriptable. Main detractor might be that artifacts might have a pretty tight expiry.. Ideally it'd not expire for at least a year.
- If nothing else works within the Github universe, we could do what OSRM's Windows workflow did until recently: push the compressed built binaries to some server (I'm happy to give one of ours), where the binding workflow can pull from.

This would be also interesting for other repos, e.g. https://github.com/VROOM-Project/pyvroom, where we'd like to be able to compile VROOM with libosrm support before binding VROOM in Python.

**Comment by whytro** [comment]
Dropping an update on the Windows side of things. Windows seems a bit more complex to manage in this regard, especially regarding the exporting of dependencies (ie. boost). Since osrm-backend doesn't use a formal conan profile, I can't use the conan `export-pkg` command, so I'll have to look into simply copying over the `.conan` folder to see if I can get that to work.

**Comment by nilsnolde** [comment]
We agreed offline to keep using conan. Likely it's just the binding's build command that's a bit off.

**Comment by whytro** [comment]
There's some symlinking going on as well, which seems to mess with the target-side, so I do need to investigate more regarding the packaging aspects of things (ie. deference flag on tar).

**Comment by whytro** [comment]
A debug build on osrm-backend with conan enabled seems to fail on the target machine (runs out of space apparently), while a vcpkg run on osrm-backend with the vcpkg toolchain enabled throws the error: `Evaluation file to be written multiple times with different content.`, and fails to configure.

**Comment by whytro** [comment]
Currently, I'm not able to get the OSRM library to be detected upon install, resulting in the Windows build failing, even though it seems like it should (ie. it's a path listed in the find debug log). I've tried pointing cmake directly at the OSRM library and headers, but it still fails to configure, and complains about missing headers.

Placing osrm-backend as a submodule yields a different result where it'll run conan, and then immediately fail to find the conan packages. osrm-backend still seems to not build with vcpkg for me as well.

**Comment by nilsnolde** [comment]
Hm, does any route/table/match test work? Do we know if it's the bindings or the OSRM code causing the segfault?

**Comment by whytro** [comment]
> Hm, does any route/table/match test work? Do we know if it's the bindings or the OSRM code causing the segfault?

Some of the other tests work, but table, match and tile tests throw an access violation, and the other tests will throw a "could not find matching segment" error for many of the tests that run without throwing an access violation. 

The cause doesn't seem to be in the bindings, since I traced it back to an osrm-backend function [here](https://github.com/Project-OSRM/osrm-backend/blob/db7946d76246cb10a5710c8f2851997417940b55/include/util/filtered_graph.hpp#L63) where it seems to stop working.

I'll try to rule out improper generation of the test data, and then run the NodeJS test suite to get a better understanding of where the problem really lies.

**Comment by nilsnolde** [comment]
Ah right, I'm not sure linux generated graphs would work with Windows. Or did you do what OSRM's `windows-build.bat` is doing and generate the graph on Windows? I'm :crossed_fingers: that it's that.

**Comment by whytro** [comment]
> Ah right, I'm not sure linux generated graphs would work with Windows. Or did you do what OSRM's `windows-build.bat` is doing and generate the graph on Windows? I'm 🤞 that it's that.

Yes, above errors were with `windows-build.bat`, since osrm-datastore will complain if using linux generated graphs and refuse to run.

**Comment by whytro** [comment]
I've confirmed that the nodejs tests seem to be running fine on the test system, while the Python bindings still seem to throw an access violation - so it seems to be something wrong with the bindings side.

**Comment by nilsnolde** [comment]
Wow congrats @whytro 🥳. That’s the icing on your project cake:)

**Comment by nilsnolde** [comment]
I'll close here. Windows is broken (again) but that's for another issue;)
